gibbon-compiler/src/Gibbon/Language/Syntax.hs

```diff
@@ -196,36 +198,44 @@ data FunMeta = FunMeta
   deriving (Read, Show, Eq, Ord, Generic, NFData, Out)
 
 -- | A function definiton indexed by a type and expression.
-data FunDef ex = FunDef { funName   :: Var
-                        , funArgs   :: [Var]
-                        , funTy     :: ArrowTy (TyOf ex)
-                        , funBody   :: ex
-                        , funMeta   :: FunMeta
-                        }
-
-deriving instance (Read ex, Read (ArrowTy (TyOf ex))) => Read (FunDef ex)
-deriving instance (Show ex, Show (ArrowTy (TyOf ex))) => Show (FunDef ex)
-deriving instance (Eq ex, Eq (ArrowTy (TyOf ex))) => Eq (FunDef ex)
-deriving instance (Ord ex, Ord (ArrowTy (TyOf ex))) => Ord (FunDef ex)
-deriving instance Generic (FunDef ex)
-deriving instance (Generic (ArrowTy (TyOf ex)), NFData ex, NFData (ArrowTy (TyOf ex))) => NFData (FunDef ex)
-deriving instance (Generic (ArrowTy (TyOf ex)), Out ex, Out (ArrowTy (TyOf ex))) =>  Out (FunDef ex)
+data FunDef loc ex = FunDef {   funName   :: Var
+                              , funArgs   :: [loc]
+                              , funTy     :: ArrowTy (TyOf ex)
+                              , funBody   :: ex
+                              , funMeta   :: FunMeta
+                            }
+
+deriving instance (Read ex, Read (ArrowTy (TyOf ex)), Read loc) => Read (FunDef loc ex)
+deriving instance (Show ex, Show (ArrowTy (TyOf ex)), Show loc) => Show (FunDef loc ex)
+deriving instance (Eq ex, Eq (ArrowTy (TyOf ex)), Eq loc) => Eq (FunDef loc ex)
+deriving instance (Ord ex, Ord (ArrowTy (TyOf ex)), Ord loc) => Ord (FunDef loc ex)
+deriving instance Generic (FunDef loc ex)
+deriving instance (Generic (ArrowTy (TyOf ex)), NFData ex, NFData (ArrowTy (TyOf ex)), NFData loc) => NFData (FunDef loc ex)
+deriving instance (Generic (ArrowTy (TyOf ex)), Out ex, Out (ArrowTy (TyOf ex)), Out loc) =>  Out (FunDef loc ex)
 
 -- | Insert a 'FunDef' into 'FunDefs'.
 -- Raise an error if a function with the same name already exists.
-insertFD :: FunDef ex -> FunDefs ex -> FunDefs ex
+insertFD :: FunDef Var ex -> FunDefs Var ex -> FunDefs Var ex
 insertFD d = M.insertWith err' (funName d) d
   where
    err' = error $ "insertFD: function definition with duplicate name: "++show (funName d)
 
+insertFD' :: FunDef LocVar ex -> FunDefs LocVar ex -> FunDefs LocVar ex
+insertFD' d = M.insertWith err' (Single $ funName d) d
+  where
+   err' = error $ "insertFD: function definition with duplicate name: "++show (funName d)
+
 -- |
-fromListFD :: [FunDef ex] -> FunDefs ex
+fromListFD :: [FunDef Var ex] -> FunDefs Var ex
 fromListFD = L.foldr insertFD M.empty
 
 -- |
-initFunEnv :: FunDefs a -> TyEnv (ArrowTy (TyOf a))
+initFunEnv :: FunDefs Var a -> TyEnv Var (ArrowTy (TyOf a))
 initFunEnv fds = M.map funTy fds
 
+initFunEnv' :: FunDefs LocVar a -> TyEnv LocVar (ArrowTy (TyOf a))
+initFunEnv' fds = M.map funTy fds
```

Comment:
Try to avoid copy-pasting. There may be enough to have a single version of `initFunEnv` and maybe `insertFD`, if you use polymorphism. E.g. does this work:

```
initFunEnv :: FunDefs v a -> TyEnv v (ArrowTy (TyOf a))
initFunEnv fds = M.map funTy fds
```
?

Reply:
Monomorphic definitions means less code change.